### PR TITLE
lib/os/heap: Correct aligned_alloc sizing for small heaps

### PR DESCRIPTION
**Pull request by andyross**
The code that made aligned_alloc work with the 4-byte heap headers was
requesting a block of the correctly padded size, and correctly
aligning the output buffer within that memory, but it was using the
UNALIGNED chunk size for the buffer as the final size of the block
with splitting off the unused suffix.  So the final chunk in the
buffer was could be incorrectly returned to the heap and reused,
leading to overlap.

Compute the chunk size of the output buffer based on the
already-aligned output pointer instead.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

### CODE REVIEW

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by npitre** [comment]
Bummer. Would be a good idea to add this case to the test suite.

I'd suggest this alternative which is IMHO more readable:

```
diff --git a/lib/os/heap.c b/lib/os/heap.c
index d00967bcad..f3a9172bd7 100644
--- a/lib/os/heap.c
+++ b/lib/os/heap.c
@@ -261,7 +261,6 @@ void *sys_heap_aligned_alloc(struct sys_heap *heap, size_t align, size_t bytes)
 	 * We over-allocate to account for alignment and then free
 	 * the extra allocations afterwards.
 	 */
-	size_t alloc_sz = bytes_to_chunksz(h, bytes);
 	size_t padded_sz = bytes_to_chunksz(h, bytes + align - 1);
 	chunkid_t c0 = alloc_chunk(h, padded_sz);
 
@@ -270,8 +269,8 @@ void *sys_heap_aligned_alloc(struct sys_heap *heap, size_t align, size_t bytes)
 	}
 
 	/* Align allocated memory */
-	void *mem = chunk_mem(h, c0);
-	mem = (void *) ROUND_UP(mem, align);
+	uint8_t *mem = chunk_mem(h, c0);
+	mem = (uint8_t *) ROUND_UP(mem, align);
 
 	/* Get corresponding chunk */
 	chunkid_t c = mem_to_chunkid(h, mem);
@@ -284,9 +283,11 @@ void *sys_heap_aligned_alloc(struct sys_heap *heap, size_t align, size_t bytes)
 	}
 
 	/* Split and free unused suffix */
-	if (chunk_size(h, c) > alloc_sz) {
-		split_chunks(h, c, c + alloc_sz);
-		free_list_add(h, c + alloc_sz);
+	uint8_t *mem_end = (uint8_t *) ROUND_UP(mem + bytes, CHUNK_UNIT);
+	chunkid_t c_tail = mem_to_chunkid(h, mem_end);
+	if (right_chunk(h, c) > c_tail) {
+		split_chunks(h, c, c_tail);
+		free_list_add(h, c_tail);
 	}
 
 	set_chunk_used(h, c, true);
```

**Comment by npitre** [comment]
On Thu, 22 Oct 2020, Nicolas Pitre wrote:

> Bummer. Would be a good idea to add this case to the test suite.

Here's the case.

```
diff --git a/tests/lib/heap_align/src/main.c b/tests/lib/heap_align/src/main.c
index d9f94fb048..fda89b8e59 100644
--- a/tests/lib/heap_align/src/main.c
+++ b/tests/lib/heap_align/src/main.c
@@ -81,6 +81,12 @@ static void test_aligned_alloc(void)
 			}
 		}
 	}
+
+	/* corner case on small heaps */
+	p = sys_heap_aligned_alloc(&heap, 8, 12);
+	memset(p, 0, 12);
+	zassert_true(sys_heap_validate(&heap), "heap invalid");
+	sys_heap_free(&heap, p);
 }
 
 void test_main(void)
```

**Comment by npitre** [comment]
On Thu, 22 Oct 2020, Nicolas Pitre wrote:

> I'd suggest this alternative which is IMHO more readable:

Screw that. I made the same mistake again.

With the test case I finally created a fix that works.
Please see PR #29460 which is actually a rework of your fix
that looks clearer (to me at least).





> 
> ```
> diff --git a/lib/os/heap.c b/lib/os/heap.c
> index d00967bcad..f3a9172bd7 100644
> --- a/lib/os/heap.c
> +++ b/lib/os/heap.c
> @@ -261,7 +261,6 @@ void *sys_heap_aligned_alloc(struct sys_heap *heap, size_t align, size_t bytes)
>  	 * We over-allocate to account for alignment and then free
>  	 * the extra allocations afterwards.
>  	 */
> -	size_t alloc_sz = bytes_to_chunksz(h, bytes);
>  	size_t padded_sz = bytes_to_chunksz(h, bytes + align - 1);
>  	chunkid_t c0 = alloc_chunk(h, padded_sz);
>  
> @@ -270,8 +269,8 @@ void *sys_heap_aligned_alloc(struct sys_heap *heap, size_t align, size_t bytes)
>  	}
>  
>  	/* Align allocated memory */
> -	void *mem = chunk_mem(h, c0);
> -	mem = (void *) ROUND_UP(mem, align);
> +	uint8_t *mem = chunk_mem(h, c0);
> +	mem = (uint8_t *) ROUND_UP(mem, align);
>  
>  	/* Get corresponding chunk */
>  	chunkid_t c = mem_to_chunkid(h, mem);
> @@ -284,9 +283,11 @@ void *sys_heap_aligned_alloc(struct sys_heap *heap, size_t align, size_t bytes)
>  	}
>  
>  	/* Split and free unused suffix */
> -	if (chunk_size(h, c) > alloc_sz) {
> -		split_chunks(h, c, c + alloc_sz);
> -		free_list_add(h, c + alloc_sz);
> +	uint8_t *mem_end = (uint8_t *) ROUND_UP(mem + bytes, CHUNK_UNIT);
> +	chunkid_t c_tail = mem_to_chunkid(h, mem_end);
> +	if (right_chunk(h, c) > c_tail) {
> +		split_chunks(h, c, c_tail);
> +		free_list_add(h, c_tail);
>  	}
>  
>  	set_chunk_used(h, c, true);
> ```
> 
>

File: lib/os/heap.c [hunk 1]
```diff
@@ -261,7 +261,6 @@ void *sys_heap_aligned_alloc(struct sys_heap *heap, size_t align, size_t bytes)
 	 * We over-allocate to account for alignment and then free
 	 * the extra allocations afterwards.
 	 */
-	size_t alloc_sz = bytes_to_chunksz(h, bytes);
 	size_t padded_sz = bytes_to_chunksz(h, bytes + align - 1);
```
Comment (lyakh) [inline lib/os/heap.c:264]: I might be misunderstanding how this works. If I want to allocate 13 bytes aligned on a 16-byte border, I call this function as `sys_heap_aligned_alloc(heap, 16, 13)`, right? So, now `padded_sz = bytes_to_chunksz(h, 29);` If it's a "small heap," we get `(29 + 4 + 7) / 8 = 5`, right? That means we want to allocate 40 bytes for 13 bytes of user data plus 4 bytes of header, if the user wants to align on a 16-byte boundary? Wouldn't 24 bytes be enough for that?

**Comment by andyross** [comment]
Close this one in favor of #29460

**Comment by npitre** [comment]
On Fri, 23 Oct 2020, Guennadi Liakhovetski wrote:

> I might be misunderstanding how this works. If I want to allocate 13 bytes aligned on a 16-byte border, I call this function as `sys_heap_aligned_alloc(heap, 16, 13)`, right? So, now `padded_sz = bytes_to_chunksz(h, 29);` If it's a "small heap," we get `(29 + 4 + 7) / 8 = 5`, right? That means we want to allocate 40 bytes for 13 bytes of user data plus 4 bytes of header, if the user wants to align on a 16-byte boundary? Wouldn't 24 bytes be enough for that?

You're right. We should subtract the header size from the padding.

```
diff --git a/lib/os/heap.c b/lib/os/heap.c
index 9f4b35ff25..adcde639be 100644
--- a/lib/os/heap.c
+++ b/lib/os/heap.c
@@ -261,7 +261,8 @@ void *sys_heap_aligned_alloc(struct sys_heap *heap, size_t align, size_t bytes)
 	 * We over-allocate to account for alignment and then free
 	 * the extra allocations afterwards.
 	 */
-	size_t padded_sz = bytes_to_chunksz(h, bytes + align - 1);
+	size_t padded_sz = bytes_to_chunksz(h, bytes + align - 1
+						- chunk_header_bytes(h));
 	chunkid_t c0 = alloc_chunk(h, padded_sz);
 
 	if (c0 == 0) {
```

I'll create a separate PR for that.

**Pull request closed without merging**